patches/src/decompression.asm

```diff
@@ -5,287 +5,630 @@
+;    this leads to a bit more code, but i believe it's worth it (there's some space left before hitting the VRAM decomp routine
+;-direct copy has been moved in front of the main loop, because that way a JMP doesn't need to get taken from there.
+;-word fill has been modified to go straight into 16-bit mode, to avoid having to load a word whilst in 8-bit mode
+;-Because storing happens more often than loading during decompression, DB is now set to the destination bank.
+;    this means loading now uses indirect long, and storing indirect short
+;    updating source bank is also simpler now
 
-++
 
-	AND #$E0
-	PHA
-	TDC
-	LDA $4A
-	AND #$1F
+;doesn't change that much, but it is faster!
 
-+++
 
-	TAX
-	INX
-	PLA
+;Overhauled by Tundain (26/12/2024: DMY)
+;Huge overhaul this time for significant performance improvement!
+; I changed the way it writes data to use the WRAM I/O port ($2180)
+; I also made the decompression commands use DMA transfers where applicable
+; I also changed the DP during decompression to $4300 and moved all temporary variables usage to over there, since DMA channel RAM is apparently faster than WRAM.
+; I also undid some changes in the previous version because they became impractical
+; and a bunch of other small tweaks
 
-	BMI Option4567
+;Also optimizes the VRAM decompression routine!
 
-	BEQ Option0
+;unfortunately now requires a bit of freespace in bank $80 
+;if any other patch you're using conflicts, feel free to repoint
 
-	CMP #$20
+!bank_80_free_space_start = $80E2A0
+!bank_80_free_space_end = $80E380
 
-	BEQ BRANCH_THETA
+!bank_89_free_space_start = $89C000
+!bank_89_free_space_end = $89C040
 
-	CMP #$40
+;also uses one unused RAM variable
+;if another patch also uses this RAM variable, find yourself a different one
+!DecompFlag = $9B
 
-	BEQ BRANCH_IOTA
+lorom
 
-;X = 3: Store an ascending number (starting with the value of the next byte) Y times.
-	LDA ($47)
-	INC $47
-	BNE +
-	JSR IncrementBank2
-+
+;;; hooks
 
--
+org $8095A7
+    JMP checkIfInLoading
 
-	STA [$4C], Y
-	INC A
-	INY
-	DEX
+org $8095FC
+    JMP checkstack
 
-	BNE -
+;change bottom of HUD hcounter, since IRQ does more work
+org $8096A5
+    LDX #$0078
 
-	JMP NextByte
+org $80972F
+    LDX #$0078
 
-Option0:
--
+;some hijacks
+org $809876
+    JMP setDp
+return:
 
-	LDA ($47)	;X = 0: Directly copy Y bytes
-	INC $47
-	BNE +
-	JSR IncrementBank2
-+
+org $809883
+    JMP ResetDp
+return2:
 
-	STA [$4C], Y
-	INY
-	DEX
+;HDMA channels aren't preserved during unpause, but since decomp uses DMA 2/3, those now need to be preserved
+org $828D1A
+    JSR preserveDMA
 
-	BNE -
+org $80A15F
+    JSR restoreDMA : NOP
 
-	JMP NextByte
+;switch 2 calls during start gameplay
+org $80A0E5
+    JSL $82E97C
+    JSL $89AB82
 
-BRANCH_THETA:
+org $82E7A8
+    JSL VRAMdecomp
 
-	LDA ($47)	;X = 1: Copy the next byte Y times.
-	INC $47
-	BNE +
-	JSR IncrementBank2
-+
+org $82E7BB
+    JSL VRAMdecomp
```

Comment:
Looks like this conflicts with the hook to load hazard markers here: https://github.com/blkerby/MapRandomizer/blob/27bf8b9ff59a2b2cbd7a6364d8743c5f05231660/patches/src/hazard_markers.asm#L16